gensim/similarities/docsim.py

```diff
@@ -563,6 +564,107 @@ def __str__(self):
         return "%s<%i docs, %i features>" % (self.__class__.__name__, len(self), self.index.shape[1])
 
 
+class SoftCosineSimilarity(interfaces.SimilarityABC):
+    """Document similarity (like MatrixSimilarity) that uses Soft Cosine Similarity as a similarity
+    measure.
+
+    Parameters
+    ----------
+    corpus: list of lists of (int, float) two-tuples
+        A list of documents in the gensim document format.
+    similarity_matrix : scipy.sparse.csc_matrix
+        A term similarity matrix.
+    num_best : int
+        The number of results to retrieve for a query.
+
+    See Also
+    --------
+    gensim.models.keyedvectors.EuclideanKeyedVectors.similarity_matrix
+        A term similarity matrix produced from term embeddings.
+    gensim.matutils.softcossim
+        The Soft Cosine Similarity.
+
+    Examples
+    --------
+    >>> from gensim.models import Word2Vec
+    >>> from gensim.similarities import SoftCosineSimilarity
+    >>> from gensim.utils import simple_preprocess
+    >>> from gensim.corpora import Dictionary
+    >>> # Given a document collection "corpus", train a word2vec model.
+    >>> model = Word2Vec(corpus, workers=3, size=100)
+    >>> # Construct a bag-of-words corpus, a dictionary, and a term similarity matrix.
+    >>> dictionary = Dictionary(corpus)
+    >>> corpus = [dictionary.doc2bow(document) for document in corpus]
+    >>> similarity_matrix = model.wv.similarity_matrix(corpus, dictionary)
+    >>> index = SoftCosineSimilarity(corpus, similarity_matrix, num_best=10)
+    >>> # Make a query.
+    >>> query = 'Yummy! Great view of the Bellagio Fountain show.'
+    >>> sims = index[dictionary.doc2bow(simple_preprocess(query))]
+
+    See `Tutorial Notebook
+    <https://github.com/RaRe-Technologies/gensim/blob/develop/docs/notebooks/soft_cosine_tutorial.ipynb>`_
+    for more examples.
+    """
+
+    def __init__(self, corpus, similarity_matrix, num_best=None, chunksize=256):
+        self.corpus = corpus
+        self.num_best = num_best
+        self.chunksize = chunksize
+
+        # Normalization of features is undesirable, since soft cosine similarity requires special
+        # normalization using the similarity matrix. Therefore, we would just be normalizing twice,
+        # increasing the numerical error.
+        self.normalize = False
+
+        # index is simply an array from 0 to size of corpus.
+        self.index = numpy.array(range(len(corpus)))
+
+        # Remove the columns of the similarity matrix that correspond to terms outside corpus.
+        nonzero_columns = sorted(set((index for document in corpus for index, _ in document)))
+        identity_matrix = scipy.sparse.identity(similarity_matrix.shape[0],
+                                                dtype=similarity_matrix.dtype, format="csr")
+        with warnings.catch_warnings():
```

Comment:
Why this block needed?

Reply:
As stated in the comment, the intent is to „remove the columns of the similarity matrix that correspond to terms outside corpus“. This will create a smaller matrix with zeroed-out columns that never appear in the corpus and that will therefore never be accessed. Due to the deficiencies in the SciPy dot product implementation, the subsequent calls to `gensim.matlibs.softcossim` will also be faster with such a matrix; an optimized dot product code would not access these columns either way.